### ISSUE TITLE
Is the 'x' sent by the client  to servers in plaintext or ciphertest?

### ISSUE THREAD

**Issue by giraffeCjl** [tf-encrypted/tf-encrypted#825]
In /tf-encrypted/tf_encrypted/serving.py, tfe defines a 'run(self, sess, x)' function which gets image from 'x_test' and sends `x` to servers. But the print result seems that 'x' is in plaintext. 
I have a question: In the homomorphic encryption computing mode, shouldn't the data be encrypted on the client and then sent to the server? Why is the input of the client in plaintext here?

**Comment by jvmncs** [comment]
**TL;DR** the inputs/outputs to this function are encrypted/decrypted for you. The QueueClient & QueueServer are working at a high-level of abstraction, and the details of how the input & output to the serving queue are kept secure is determined by the current protocol (& config) being used.

-----

I think you're referring to this function:
https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/serving/queues.py#L140-L145

This calls `QueueClient.send_input`, which is probably the source of confusion:
https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/serving/queues.py#L112-L125

The name `send_input` is a bit misleading here, especially given that it's checking the input to be a raw numpy array.  However, there is work being done behind the scenes to make this numpy array secret before it's sent away from the `QueueClient`. This extra work is defined here, when we call `tfe.define_private_placeholder`:
https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/serving/queues.py#L87-L93

Under the hood, this is creating a `tf.Placeholder` that is then secret shared according to the currently active protocol, producing a reference to some future private value. This secret sharing happens wherever the `QueueClient` is being used, so the result of `send_input` is actually a set of multiple shares making up the private value. The `QueueServer` is then a wrapper around a protocol-specific Queue, and for Pond this is essentially managing one `FIFOQueue` for each compute server's physical device. The shares in this queue can then be used in further protocol computation before being dequeued and returned to the client, where the shares will be reconstructed into the plaintext computation result (which is what the client sees at the end of it all).

**Comment by giraffeCjl** [comment]
> **TL;DR** the inputs/outputs to this function are encrypted/decrypted for you. The QueueClient & QueueServer are working at a high-level of abstraction, and the details of how the input & output to the serving queue are kept secure is determined by the current protocol (& config) being used.
> 
> I think you're referring to this function:
> 
> https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/serving/queues.py#L140-L145
> 
> This calls `QueueClient.send_input`, which is probably the source of confusion:
> 
> https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/serving/queues.py#L112-L125
> 
> The name `send_input` is a bit misleading here, especially given that it's checking the input to be a raw numpy array. However, there is work being done behind the scenes to make this numpy array secret before it's sent away from the `QueueClient`. This extra work is defined here, when we call `tfe.define_private_placeholder`:
> 
> https://github.com/tf-encrypted/tf-encrypted/blob/8b7cfb32c426e9a6f56769a1b47626bd1be03a66/tf_encrypted/serving/queues.py#L87-L93
> 
> Under the hood, this is creating a `tf.Placeholder` that is then secret shared according to the currently active protocol, producing a reference to some future private value. This secret sharing happens wherever the `QueueClient` is being used, so the result of `send_input` is actually a set of multiple shares making up the private value. The `QueueServer` is then a wrapper around a protocol-specific Queue, and for Pond this is essentially managing one `FIFOQueue` for each compute server's physical device. The shares in this queue can then be used in further protocol computation before being dequeued and returned to the client, where the shares will be reconstructed into the plaintext computation result (which is what the client sees at the end of it all).
Thank you for such a detailed answer, the send_input function is indeed a bit misleading, and I studied it until late lol.
In general, can I understand this way, the QueueClient can see the calculated ciphertext result and the decrypted output, but the QueueServer cannot output the ciphertext in the whole process? I once tried to only output a ‘PondPrivateTensor’ on the QueueServer side.

**Comment by giraffeCjl** [comment]
@jvmncs Thank you for such a detailed answer, the send_input function is indeed a bit misleading, and I studied it until late LOL.
In general, can I understand this way, the QueueClient can see the calculated ciphertext result and the decrypted output, but the QueueServer cannot output the ciphertext in the whole process? I once tried to only output a ‘PondPrivateTensor’ on the QueueServer side.

**Comment by zicofish** [comment]
Closing this. Please open a new issue if there is further question.
